websearch/websearch/src/durability.rs

```diff
@@ -0,0 +1,579 @@
+                    }
+                });
+
+                let persisted_params = result?;
+                durability.persist_infallible(
+                    StartSearchInput {
+                        params: persisted_params.clone(),
+                    },
+                    persisted_params.clone(),
+                );
+                Ok(SearchSession::new(DurableSearchSession::<Impl>::live(
+                    Impl::unwrapped_search_session(persisted_params).unwrap(),
+                )))
+            } else {
+                let result = durability.replay_infallible();
+                let session = SearchSession::new(DurableSearchSession::<Impl>::replay(result));
+                Ok(session)
+            }
+        }
+        fn search_once(
+            params: SearchParams,
+        ) -> Result<(Vec<SearchResult>, Option<SearchMetadata>), SearchError> {
+            let durability =
+                Durability::<(Vec<SearchResult>, Option<SearchMetadata>), UnusedError>::new(
+                    "golem_websearch",
+                    "search_once",
+                    DurableFunctionType::WriteRemote,
+                );
+            if durability.is_live() {
+                let result = with_persistence_level(PersistenceLevel::PersistNothing, || {
+                    Impl::search_once(params.clone())
+                });
+                let (results, metadata) = result?;
+                durability.persist_infallible(
+                    SearchOnceInput { params },
+                    (results.clone(), metadata.clone()),
+                );
+                Ok((results, metadata))
+            } else {
+                let result: (Vec<SearchResult>, Option<SearchMetadata>) =
+                    durability.replay_infallible();
+                let (results, metadata) = result;
+                Ok((results, metadata))
+            }
+        }
+    }
+
+    /// Represents the durable search session's state
+    ///
+    /// In live mode it directly calls the underlying websearch session which is implemented on
+    /// top of HTTP requests to search providers.
+    ///
+    /// In replay mode it buffers the replayed search results, and also tracks the created pollables
+    /// to be able to reattach them to the new live session when the switch to live mode
+    /// happens.
+    ///
+    /// When reaching the end of the replay mode, if the replayed session was not finished yet,
+    /// the retry parameters implemented in `ExtendedwebsearchGuest` is used to create a new websearch session
+    /// and continue the search seamlessly.
+    enum DurableSearchSessionState<Impl: ExtendedwebsearchGuest> {
+        Live {
+            session: Impl::SearchSession,
+            pollables: Vec<LazyInitializedPollable>,
+        },
+        Replay {
+            original_params: SearchParams,
+            pollables: Vec<LazyInitializedPollable>,
+            partial_results: Vec<SearchResult>,
```

Comment:
This looks incorrect. From what I can tell you are never writing to the partial results.